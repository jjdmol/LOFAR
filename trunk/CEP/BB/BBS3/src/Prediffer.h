//# Prediffer.h: Read and predict read visibilities
//#
//# Copyright (C) 2004
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id$

#ifndef BB_BBS3_PREDIFFER_H
#define BB_BBS3_PREDIFFER_H

#include <casa/Arrays/Matrix.h>
#include <casa/Quanta/MVBaseline.h>

#include <BBS3/ParmData.h>
#include <BBS3/MNS/MeqDomain.h>
#include <BBS3/MNS/MeqHist.h>
#include <BBS3/MNS/MeqJonesExpr.h>
#include <BBS3/MNS/MeqJonesNode.h>
#include <BBS3/MNS/MeqMatrix.h>
#include <BBS3/MNS/MeqParm.h>
#include <BBS3/MNS/MeqPhaseRef.h>
#include <BBS3/MNS/MeqSourceList.h>
#include <BBS3/MNS/MeqRequest.h>
#include <BBS3/MNS/MeqStation.h>
#include <BBS3/MNS/MeqStatSources.h>
#include <BBS3/MNS/MeqLofarStatSources.h>
#include <BBS3/MNS/MeqStatUVW.h>
#include <BBS3/MNS/ParmTable.h>

#include <Common/Timer.h>
#include <Common/LofarTypes.h>
#include <Common/lofar_string.h>
#include <list>

namespace LOFAR
{

//# Forward Declarations
class MMap;

// Prediffer calculates the equations for the solver.
// It reads the measured data and predicts the data from the model.
// It subtracts them from each other and calculates the derivatives.
// These results can be sent to the solver to find better parameter values.

class Prediffer
{

public:
  // Create Prediffer object for a specific
  // MeaurementSet, MEQ model (with associated MEP database) and skymodel
  // for the specified data descriptor (i.e. spectral window) and antennas.
  // The database type (aips or postgres) has to be given.
  // For postgres the database name has to be given as well.
  // Currently model types WSRT, LOFAR.RI and LOFAR.AP are recognized.
  // The UVW coordinates can be recalculated or taken from the MS.
  Prediffer (const string& msName,
	     const string& meqModel,
	     const string& skyModel,
	     const string& dbType,
	     const string& dbName,
	     const string& dbHost,
	     const string& dbPwd,
	     const vector<int>& ant,
	     const string& modelType,
	     bool calcUVW,
	     bool lockMappedMem);

  // Destructor
  ~Prediffer();

  // Make a selection of the MS to be used in the domain iteration.
  void select (const vector<int>& ant1, const vector<int>& ant2,
	       bool useAutoCorrelations);

  // Set the domain (in frequency and time).
  // Hereafter getSolvableParmData can be called.
  // It returns a vector containing the 3-dim shape of the expected data
  // (as nrchan-nrspid-nreq).
  // The 4th element of the vector is the total number of equations.
  // The vector is empty if the domain is outside the observation domain.
  // Length is trimmed if beyond end of observation.
  vector<uint32> setDomain (double startFreq, double lengthFreq,
			    double startTime, double lengthTime);

  // Update the solvable parm values (reread from table).
  void updateSolvableParms();

  // Return the solvable parms.
  // The parms are in ascending order of spidnr.
  const vector<ParmData>& getSolvableParmData() const
    { return itsParmData; }

  // Make all parameters non-solvable.
  void clearSolvableParms();

  // Make specific parameters solvable (isSolvable = True) or
  // non-solvable (False).
  void setSolvableParms (const vector<string>& parms, 
			 const vector<string>& excludePatterns,
			 bool isSolvable);

  // Get the equations for all selected baselines.
  // The values are stored into the buffer as a 3-dim array with axes
  // nresult,nspid+1,nval.
  // It is checked if the shape of the buffer is correct.
  // The function should be called until a false status is returned.
  // In nresult it returns the number of results put in the buffer.
  // Normally this is shape(2), but for the last time it can be less.
  bool getEquations (double* result, const vector<uint32>& shape,
		     int& nresult);

  // Set the source numbers to use in this peel step.
  bool setPeelSources (const vector<int>& peelSources,
		       const vector<int>& extraSources);

  // Subtract the peel source(s) from the data.
  // If write=true, the data are written into a new file.
  void subtractPeelSources (bool write=false);

  // Set the names and values of all solvable parms for the current domain.
  // The double version can only be used if all parms are 0th-order
  // polynomials.
  // <group>
  void getParmValues (vector<string>& names,
		      vector<double>& values);
  void getParmValues (vector<string>& names,
		      vector<MeqMatrix>& values);
  // </group>

  // Set the given values (for the current domain) of parms matching
  // the corresponding name.
  // Values with a name matching no parm, are ignored.
  // <group>
  void setParmValues (const vector<string>& names,
		      const vector<double>& values);
  void setParmValues (const vector<string>& names,
		      const vector<MeqMatrix>& values);
  // </group>

  // Show the settings of the Prediffer.
  void showSettings() const;

private:
  // Copy constructor and assignment are not allowed.
  // <group>
  Prediffer (const Prediffer& other);
  Prediffer& operator= (const Prediffer& other);
  // </group>

  // Get measurement set description from file
  void readDescriptiveData (const string& fileName);

  // initialize all parameters in the MeqExpr tree for the current domain
  void initParms (const MeqDomain& domain, bool callReadPocs);

  // Get the phase reference position of the first field.
  void getPhaseRef (double ra, double dec, double startTime);

  // Get the station info (position and name).
  void fillStations (const vector<int>& antnrs);

  // Get all baseline info.
  void fillBaselines (const vector<int>& antnrs);

  // Count nr of baselines selected.
  void countBaselines();

  // Fill all UVW coordinates if they are not calculated.
  void fillUVW();

  // Create the WSRT expressions for each baseline.
  void makeWSRTExpr();

  // Create the LOFAR expressions for each baseline.
  // The EJones can be expressed as real/imag or ampl/phase.
  void makeLOFARExpr (casa::Bool asAP);

  // Get equations for a single time and baseline.
  void getEquation (double* result, const fcomplex* data,
		    const MeqRequest& request,
		    int blindex, int ant1, int ant2);


  string                itsMSName;      //# Measurement set name
  string                itsMEPName;     //# Common parmtable name
  ParmTable             itsMEP;         //# Common parmtable
  string                itsGSMMEPName;  //# GSM parameters parmtable name
  ParmTable             itsGSMMEP;      //# parmtable for GSM parameters
  MeqParmGroup          itsParmGroup;   //# container for all parms
  bool                  itsCalcUVW;

  MeqPhaseRef           itsPhaseRef;    //# Phase reference position in J2000

  MeqSourceList         itsSources;
  casa::Vector<int>     itsPeelSourceNrs;
  vector<MeqStation*>   itsStations;
  vector<MeqStatUVW*>   itsStatUVW;
  vector<MeqStatSources*> itsStatSrc;
  vector<MeqLofarStatSources*> itsLSSExpr; //# Lofar sources per station
  vector<MeqJonesExpr*> itsStatExpr;    //# Expression per station
  vector<MeqExpr*>      itsComplexConv; //# Expression to convert to complex
  vector<MeqJonesNode*> itsJonesNodes;  //# Jones nodes
  vector<casa::MVBaseline>     itsBaselines;
  vector<MeqHist>       itsCelltHist;   //# Histogram of #cells in time
  vector<MeqHist>       itsCellfHist;   //# Histogram of #cells in freq
  vector<MeqJonesExpr*> itsExpr;        //# solve expression tree per baseline
  vector<MeqJonesExpr*> itsResExpr;     //# residual expr tree per baseline

  double itsStartFreq;                //# start frequency of observation
  double itsEndFreq;
  double itsStepFreq;
  int    itsNrChan;                   //# nr of channels in observation
  int    itsFirstChan;                //# first channel of selected domain
  int    itsLastChan;                 //# last channel of selected domain

  string itsSolveFileName;            //# Data file used (.dat or .res)

  int          itsNrScid;             //# Nr of solvable parameter coeff.
  vector<bool> itsIsParmSolvable;     //# is corresponding parmlist solvable?
  vector<ParmData> itsParmData;       //# solvable parm info. 

  vector<int>          itsAnt1;        //# Antenna1 antenna numbers
  vector<int>          itsAnt2;        //# Antenna2 antenna numbers
  int                  itsNPol;        //# Number of polarisations
  casa::Vector<double> itsTimes;       //# All times in MS
  casa::Vector<double> itsIntervals;   //# All intervals in MS
  casa::Matrix<double> itsAntPos;      //# All antenna positions
  unsigned int         itsNrBl;        //# Total number of unique baselines
  casa::Matrix<bool>   itsBLSelection; //# true = baseline is selected
  casa::Matrix<int>    itsBLIndex;     //# baseline index of antenna pair
                                       //# -1 is baseline is absent
  unsigned int   itsNrSelBl;       //# nr of selected baselines
  unsigned int   itsTimeIndex;     //# The index of the current time
  unsigned int   itsNrTimes;       //# The number of times in the time domain
  unsigned int   itsNrBufTB;       //# Nr of times/baselines fitting in buffer
  unsigned int   itsNrTimesDone;   //# The number of times done in time domain
  unsigned int   itsBlNext;        //# Next baseline to do in time domain

  MMap*          itsDataMap;       //# Data file to map
  bool           itsLockMappedMem; //# Lock memory immediately after mapping?

  NSTimer itsPredTimer;
  NSTimer itsEqTimer;

};

} // namespace LOFAR

#endif
