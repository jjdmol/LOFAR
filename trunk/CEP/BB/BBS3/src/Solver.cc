//# Solver.cc: Calculate parameter values using a least squares fitter
//#
//# Copyright (C) 2004
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id$

#include <lofar_config.h>

#include <BBS3/Solver.h>
#include <Common/VectorUtil.h>
#include <Common/LofarLogger.h>
#include <Common/Timer.h>

#include <iostream>
#include <fstream>
#include <stdexcept>


namespace LOFAR
{

//----------------------------------------------------------------------
//
// ~Solver
//
// Constructor. Initialize a Solver object.
//
// Create list of stations and list of baselines from MS.
// Create the MeqExpr tree for the WSRT.
//
//----------------------------------------------------------------------
Solver::Solver ()
: itsSolver (1),
  itsDoSet  (true)
{
  LOG_INFO_STR( "Solver constructor" );
}

//----------------------------------------------------------------------
//
// ~~Solver
//
// Destructor for a Solver object.
//
//----------------------------------------------------------------------
Solver::~Solver()
{
  LOG_TRACE_FLOW( "Solver destructor" );
}


// ~solve
//
// Solve for the solvable parameters on the current time domain.
// Returns solution and fitness.
//
//----------------------------------------------------------------------

void Solver::solve (bool useSVD,
		    Quality& resultQuality)
{
  LOG_INFO_STR( "solve using file ");

  NSTimer timer;
  timer.start();
  ASSERT (itsSolvableValues.size() > 0);

  // Initialize the solver (needed after a setSolvable).
  // Note it is usually already done in setEquations.
  if (itsDoSet) {
    itsSolver.set (itsSolvableValues.size());
    itsDoSet = false;
  }
  // Solve the equation. 
  uint rank;
  double fit;
  LOG_INFO_STR( "Solution before: " << itsSolvableValues);
  bool solFlag = itsSolver.solveLoop (fit, rank, &(itsSolvableValues[0]),
				      useSVD);
  LOG_INFO_STR( "Solution after:  " << itsSolvableValues);

  resultQuality.init();
  resultQuality.itsSolFlag = solFlag;
  resultQuality.itsRank = rank;
  resultQuality.itsFit = fit;
  resultQuality.itsMu = itsSolver.getWeightedSD();
  resultQuality.itsStddev = itsSolver.getSD();
  resultQuality.itsChi = itsSolver.getChi();
  cout << resultQuality << endl;

  // Store the new values in the ParmData vector.
  const double* val = &(itsSolvableValues[0]);
  for (uint i=0; i<itsSolvableParms.size(); ++i) {
    ParmData& parm = itsSolvableParms[i];
    double* value = parm.getRWValues().doubleStorage();
    for (int j=0; j<parm.getNrSpid(); ++j) {
      value[j] = *val++;
    }
  }

  timer.stop();
  cout << "BBSTest: solver     " << timer << endl;
  return;
}


void Solver::setEquations (const double* data, int nresult, int nrspid,
			   int nval, int prediffer)
{
  ASSERT (uint(prediffer) < itsIndices.size());
  // Initialize the solver (needed after a setSolvable).
  if (itsDoSet) {
    itsSolver.set (itsSolvableValues.size());
    itsDoSet = false;
  }
  vector<int>& predInx = itsIndices[prediffer];
  ASSERT (uint(nrspid) == predInx.size());
  // Use a consecutive vector to assemble all derivatives.
  vector<double> derivVec(nrspid);
  double* derivs = &(derivVec[0]);
  // Each result is a 2d array of [nrval,nrspid+1] (nrval varies most rapidly).
  // The first value is the difference; the others the derivatives. 
  const double* ddata = data;
  int nrval = nval;
  for (int i=0; i<nresult; ++i) {
    for (int j=0; j<nrval; ++j) {
      // Each value result,time,freq gives an equation.
      double diff = ddata[0];
      const double* derivdata = ddata + nrval;
      for (int k=0; k<nrspid; ++k) {
	derivs[k] = derivdata[k*nrval];
      }
      itsSolver.makeNorm (predInx.size(), &(predInx[0]), &(derivVec[0]),
      			  1., diff);
      // Go to next time,freq value.
      ddata++;
    }
    // Go to next result (note that ddata has already been increment nrval
    // time, so here we use nrspid instead of nrspid+1.
    ddata += nrspid*nrval;
  }
}

void Solver::initSolvableParmData (int nrPrediffers)
{
  itsSolvableMap.clear();
  itsSolvableParms.resize (0);
  itsIndices.resize (nrPrediffers);
  itsSolvableValues.resize (0);
  itsDoSet = true;
}

// The solvable parameters coming from the prediffers have to be collected
// and put into a single solvercollection.
// It means that the prediffer spids have to be mapped to a solver spids.
void Solver::setSolvableParmData (const std::vector<ParmData>& data,
				  int prediffer)
{
  if (data.size() == 0) {
    return;
  }
  itsDoSet = true;
  // Usually the last entry has the highest spid; so use that to reserve to
  // avoid (hopefully) resizes.
  vector<int> predInx;
  predInx.reserve (data[data.size()-1].getLastSpid() + 1);
  // Loop through all parm entries and see if already used.
  for (uint i=0; i<data.size(); ++i) {
    int parminx = itsSolvableMap.size();
    int spidinx = itsSolvableValues.size();
    ParmData parm(data[i]);
    map<string,int>::const_iterator value =
                                     itsSolvableMap.find (parm.getName());
    if (value == itsSolvableMap.end()) {
      // Not in use yet; so add it to the map.
      itsSolvableMap[parm.getName()] = parminx;
      // Its first solver spid is the length of the vector so far.
      parm.setFirstSpid (spidinx);
      // Save the parm.
      itsSolvableParms.push_back (parm);
      // Store the current values in the vector.
      const double* value = parm.getValues().doubleStorage();
      for (int i=0; i<parm.getNrSpid(); ++i) {
	itsSolvableValues.push_back (value[i]);
      }
    } else {
      // The parameter has already been used. Get its index.
      parminx = value->second;
      ASSERT (itsSolvableParms[parminx] == parm);
      spidinx = itsSolvableParms[parminx].getFirstSpid();
    }
    // Store the solver spid to get a mapping of prediffer spid to solver spid.
    int spidnr = parm.getFirstSpid();
    for (int i=0; i<parm.getNrSpid(); ++i) {
      predInx.push_back (spidnr++);
    }
  }
  ASSERT (uint(prediffer) < itsIndices.size());
  itsIndices[prediffer] = predInx;
}

} // namespace LOFAR

//# Instantiate the makeNorm template.
#include <scimath/Fitting/LSQFit2.cc>
template void casa::LSQFit::makeNorm<double, double*, int*>(unsigned, int* const&, double* const&, double const&, double const&, bool, bool);
