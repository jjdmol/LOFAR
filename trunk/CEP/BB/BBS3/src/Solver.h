//# Solver.h: Calculate parameter values using a least squares fitter
//#
//# Copyright (C) 2004
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id$

#ifndef LOFAR_BB_BBS3_SOLVER_H
#define LOFAR_BB_BBS3_SOLVER_H

// \file Solver.h
// Calculates parameter values using a least squares fitter

#include <scimath/Fitting/LSQaips.h>
#include <BBS3/ParmData.h>
#include <BBS3/Quality.h>
#include <Common/LofarTypes.h>
#include <Common/lofar_vector.h>
#include <Common/lofar_string.h>
#include <Common/lofar_map.h>

namespace LOFAR
{

// \addtogroup BBS3
// @{

// Solver calculates new parameter values from the equations given by the
// Prediffer class.
class Solver
{

public:
  // Create Solver object.
  Solver();

  // Destructor
  ~Solver();

  // Initialize the solvable parm data.
  void initSolvableParmData (int nrPrediffers);

  // Set the solvable parm data for a given prediffer.
  void setSolvableParmData (const vector<ParmData>&, int prediffer);

  // Get the solvable parameter name info.
  const vector<ParmData>& getSolvableParmData() const
    { return itsSolvableParms; }

  // Get the solvable parameter values.
  const vector<double>& getSolvableValues() const
    { return itsSolvableValues; }

  // Set the equations for a given prediffer.
  // The data array has to be 4-dimensional with C-style shape
  // [nresult,nrspid+1,nval].
  // The first value on the spid axis is the difference between measured and
  // predicted data. The other values are the derivatives for each spid.
  // An equation is added to the solver for each value,result.
  // Note that nrspid has to match the ParmData object given to
  // setSolvableParmData for this prediffer.
  // Also note that setEquations cannot be called before setSolvableParmData
  // has been called for all prediffers.
  // After the last setEquations, the solve function can be called.
  void setEquations (const double* data, int nresult, int nrspid,
		     int nval, int prediffer);

  // Solve which returns solved parameter values in a vector and fit value 
  // in Quality object.
  void solve (bool useSVD,
	      Quality& resultQuality);

  // Get the solutions per prediffer.
  vector<vector<double> > getSolutions() const;

private:
  // Copy constructor and assignment are not allowed.
  // <group>
  Solver(const Solver& other);
  Solver& operator=(const Solver& other);
  // </group>


  casa::LSQaips  itsSolver;
  MeqMatrix      itsSolution;            //# Solution as complex numbers
  vector<double> itsFitME;
  Quality itsSol;                        //# Solution quality
  map<string,int>      itsSolvableMap;   //# Map solv.parameter name to index
  vector<ParmData>     itsSolvableParms; //# All solvable parameters
  vector<vector<int> > itsIndices;       //# Index of coefficient in prediffer
                                         //# to coefficient in solver
  vector<double>       itsSolvableValues;
  bool                 itsDoSet;         //# true = do itsSolver.set
};

// @}

} // namespace LOFAR

#endif
