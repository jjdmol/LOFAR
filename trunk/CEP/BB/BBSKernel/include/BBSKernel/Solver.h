//# Solver.h: Calculate parameter values using a least squares fitter
//#
//# Copyright (C) 2004
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id$

#ifndef LOFAR_BB_BBS_SOLVER_H
#define LOFAR_BB_BBS_SOLVER_H

// \file
// Calculates parameter values using a least squares fitter

#include <scimath/Fitting/LSQaips.h>
#include <BBSKernel/ParmData.h>
#include <BBSKernel/Quality.h>
#include <ParmDB/ParmDB.h>
#include <Common/LofarTypes.h>
#include <Common/lofar_vector.h>
#include <Common/lofar_string.h>
#include <Common/lofar_map.h>

namespace LOFAR
{
namespace BBS
{

// \addtogroup BBS
// @{

// Solver calculates new parameter values from the equations given by the
// Prediffer class.
class Solver
{

public:
  // Create Solver object.
  Solver();

  // Destructor
  ~Solver();

  // Initialize the solvable parm data.
  void initSolvableParmData (int nrPrediffers,
			     const vector<MeqDomain>& solveDomains,
			     const MeqDomain& workDomain);

  // Set the solvable parm data for a given prediffer.
  void setSolvableParmData (const ParmDataInfo&, int prediffer);

  // Get the solvable parameter name info.
  const ParmDataInfo& getSolvableParmData() const
    { return itsParmInfo; }

  // Merge the fitters from the given Prediffer with the global fitter.
  void mergeFitters (const vector<casa::LSQFit>&, int prediffer);

  // Solve all the matrices.
  void solve (bool useSVD);

  // Get the solvable values for the given fitter (i.e. solve domain).
  const vector<double>& getSolvableValues (uint fitterIndex) const;

  // Get quality for the given fitter (i.e. solve domain).
  const Quality& getQuality (uint fitterIndex) const;

  // Show the relevant info.
  void show (std::ostream&);
  
  // Log the quality indicators, and the coefficients of the solvable
  // parameters to a ParmDB.
  void log(ParmDB::ParmDB &table);

private:
  // Copy constructor and assignment are not allowed.
  // <group>
  Solver(const Solver& other);
  Solver& operator=(const Solver& other);
  // </group>

  // The solver combines the results from multiple Prediffers.
  // Each Prediffer casn have a subset of the solvable parms and a
  // subset of the solve domains.
  // setSolvableData tells for each Prediffer its solvable parms and
  // its solve domains (as contained in ParmDataInfo). For each parm,
  // ParmData gives the number of solvable coefficients per solve domain
  // (in ParmData::PDInfo) and the number of perturbed values (which is the
  // maximum of the number of solvable coefficients for all solve domains).
  //
  // So Solver has to combine the solvable parm info from all Prediffers:
  // - It has to put the LSQFit info from the Prediffers at the correct
  //   place in the global LSQFit objects.
  // - The parm coefficients in the Prediffers have to updated correctly.
  // - The parms in the ParmDB have to updated correctly.
  // For this purpose it holds a few maps per Prediffer:
  // - index of the i-th Prediffer parm in the global parm vector.
  // - per parm a vector telling the number of scids per solve domain.
  // - per parm a vector telling the first scid in the LSQFit per solve domain.
  struct FitterData
  {
    casa::LSQFit   fitter;
    Quality        quality;
    std::vector<double> solvableValues;
    int            nused;
    int            nflagged;
  };
  
  struct PredifferInfo
  {
    vector<int>           solveDomainIndices;
    vector<vector<uint> > scidMap; //# map scids in Prediffer fitters to global
  };

  map<string,int>       itsNameMap;        //# Map parameter name to index
  vector<FitterData>    itsFitters;
  ParmDataInfo          itsParmInfo;       //# Global parm info
  vector<PredifferInfo> itsPredInfo;       //# prediffer info
  bool                  itsDoSet;          //# true = do itsSolver.set
};

// @}

} // namespace BBS
} // namespace LOFAR

#endif
