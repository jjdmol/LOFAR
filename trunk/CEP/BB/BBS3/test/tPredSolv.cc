//# tPredSolv.cc: Test program for Prediffer and Solver classes
//#
//# Copyright (C) 2005
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id$

#include <BBS3/Prediffer.h>
#include <BBS3/Solver.h>
#include <Common/VectorUtil.h>
#include <Common/LofarLogger.h>
#include <stdexcept>
#include <iostream>
#include <iomanip>
#include <malloc.h>

using namespace LOFAR;
using namespace std;

// Note:
// demo3.MS contains 50 frequency channels of 500000 Hz with
// center frequencies of 137750000-162250000 Hz.
// There are 5 time stamps of 2 sec in it (centers 2.35208883e9 + 2-10).

void doSolve (Prediffer& pre1, const vector<string>& solv)
{
  // Set the solvable parameters.
  pre1.clearSolvableParms();
  pre1.setSolvableParms (solv, vector<string>(), true);
  // Set a domain.
  ///  vector<uint32> shape = pre1.setDomain (137750000-250000, 2*500000,
  vector<uint32> shape = pre1.setDomain (0., 1e12,
					 0., 1e12);
  uint nrval = shape[0] * shape[1] * shape[2];
  uint bufnreq = shape[2];
  uint totnreq = shape[3];
  uint nrloop = (totnreq + bufnreq - 1) / bufnreq;
  cout << "bufShape " << shape << endl;
  double* buffer = new double[nrval];
    
  // Get the ParmData from the Prediffer and send it to the solver.
  Solver solver;
  solver.initSolvableParmData (1);
  solver.setSolvableParmData (pre1.getSolvableParmData(), 0);

  // Get the equations from the prediffer and give them to the solver.
  for (uint i=0; i<nrloop; i++) {
    bool more = pre1.getEquations (buffer, shape);
    uint nreq = bufnreq;
    if (i == nrloop-1) {
      nreq = totnreq  - (nrloop-1)*bufnreq;
      ASSERT (!more);
    } else {
      ASSERT (more);
    }
    solver.setEquations (buffer, nreq, shape[1], shape[0], 0);
  }

  // Do the solve.
  streamsize prec = cout.precision();
  cout << "Before: " << setprecision(10) << solver.getSolvableValues() << endl;
  Quality quality;
  solver.solve (false, quality);
  cout << "After:  " << setprecision(10) << solver.getSolvableValues() << endl;
  cout.precision (prec);
  delete [] buffer;
}


int main (int argc, const char* argv[])
{
  cout << ">>>" << endl;
  try {
    if (argc < 4) {
      cerr << "Run as: tPredSolv msname meqparmtable skyparmtable" << endl;
      return 1;
    }
    // Do a solve for RA using a few stations.
    {
      vector<int> antVec(10);
      for (uint i=0; i<antVec.size(); ++i) {
	antVec[i] = 2*i;
      }
      Prediffer pre1(argv[1], argv[2], argv[3], "aips", "test", "", "",
		     antVec, "LOFAR.RI", false, true);
      
      // Do a further selection of a few stations.
      vector<int> antVec2(10);
      for (uint i=0; i<antVec2.size(); ++i) {
	antVec2[i] = 4*i;
      }
      pre1.select (antVec2, antVec2, false);    // no autocorrelations
      vector<string> solv(3);
      solv[0] = "RA.*";
      solv[1] = "DEC.*";
      solv[2] = "StokesI.*";
      doSolve (pre1, solv);
    }
    // Do a solve for all position info using all stations.
    {
      vector<int> antVec(100);
      for (uint i=0; i<antVec.size(); ++i) {
	antVec[i] = i;
      }
      Prediffer pre1(argv[1], argv[2], argv[3], "aips", "test", "", "",
		     antVec, "LOFAR.RI", false, true);
      vector<string> solv(3);
      solv[0] = "RA.*";
      solv[1] = "DEC.*";
      solv[2] = "StokesI.*";
      doSolve (pre1, solv);
    }
  } catch (std::exception& x) {
    cerr << "Unexpected exception: " << x.what() << endl;
    return 1;
  }
  cout << "<<<" << endl;
  return 0;
}
