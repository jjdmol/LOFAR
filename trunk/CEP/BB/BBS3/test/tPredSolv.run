#!/bin/sh

\rm -rf /tmp/test.demo*
a_wd=`pwd`
ln -s /data/PSS3/diepen/BBS2/demo3 /tmp/test.demo
ln -s $a_wd/diepen.demo3-1.MEP /tmp/test.demo.MEP
ln -s $a_wd/diepen.demo3_gsm-1.MEP /tmp/test.demo_gsm.MEP

./tPredSolv /tmp/test.demo/demo demo demo_gsm
