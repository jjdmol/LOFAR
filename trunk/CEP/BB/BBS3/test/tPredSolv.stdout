>>>
log4cplus:ERROR No appenders could be found for logger (CEP.BB.BBS3).
log4cplus:ERROR Please initialize the log4cplus system properly.
/tmp/test.demo/demo
Mon Jan 31 10:46:49 2005      WARN MeasIERS::fillMeas(MeasIERS::Files, Double) (file /lofarbuild/aips++/stable/code/measures/implement/Measures/MeasIERS.cc, line 94):
Requested JD27223.3 is outside the IERS table data range
Calculations will proceed with less precision

Mon Jan 31 10:46:49 2005      WARN MeasTable::dUT1(Double) (file /lofarbuild/aips++/stable/code/measures/implement/Measures/MeasTable.cc, line 6437):
No requested dUT1 data available from IERS tables. 
Proceeding with probably less precision.

Mon Jan 31 10:46:49 2005      WARN MeasTable::PolarMotion(Double) (file /lofarbuild/aips++/stable/code/measures/implement/Measures/MeasTable.cc, line 6265):
No requested polar motion data available from IERS tables. 
Proceeding with probably less precision.

BBSTest: BeginOfInterval
BBSTest: file-mapping timer: avg = 71.6 us, total = 71.6 us, count =         1

BBSTest: initparms    timer: avg = 32.9 ms, total = 32.9 ms, count =         1

bufShape [100,10,12,200]
BBSTest: predict timer: avg = 1.81 ms, total = 90.3 ms, count =        50

BBSTest: formeqs timer: avg = 72.6 us, total = 3.63 ms, count =        50

Before: [2.734003,0.4537895,1,2.7340205,0.45368975,1,2.733981,0.45375025,1]
itsSolFlag: 1
itsRank   : 9
itsFit    : 1
itsMu     : 0.01097264283
itsStddev : 0.01097264283
itsChi    : 2.406894223

BBSTest: solver     timer: avg =  280 us, total =  280 us, count =         1

After:  [2.734004034,0.453799903,0.60459961,2.734020143,0.4536931581,0.5967461872,2.733981712,0.4537585098,0.6003633806]
/tmp/test.demo/demo
BBSTest: BeginOfInterval
BBSTest: file-mapping timer: avg =  106 us, total =  106 us, count =         1

BBSTest: initparms    timer: avg =  296 ms, total =  296 ms, count =         1

bufShape [100,10,12,99000]
BBSTest: predict timer: avg = 1.84 ms, total = 45.5  s, count =     24750

BBSTest: formeqs timer: avg = 58.7 us, total = 1.45  s, count =     24750

Before: [2.734003,0.4537895,1,2.7340205,0.45368975,1,2.733981,0.45375025,1]
itsSolFlag: 1
itsRank   : 9
itsFit    : 1
itsMu     : 0.1513780999
itsStddev : 0.1513780999
itsChi    : 226861.5521

BBSTest: solver     timer: avg =  270 us, total =  270 us, count =         1

After:  [2.734000947,0.4537899633,0.8540249608,2.734020251,0.4536898586,0.5032648719,2.733980648,0.4537503535,0.3403978668]
<<<
